crates/metaslang/cst/src/text_index.rs

```diff
@@ -39,75 +63,48 @@ impl Display for TextIndex {
 
 impl From<&str> for TextIndex {
     fn from(s: &str) -> Self {
-        let mut utf8 = 0;
-        let mut utf16 = 0;
-        let mut char = 0;
-        for c in s.chars() {
-            utf8 += c.len_utf8();
-            utf16 += c.len_utf16();
-            char += 1;
+        let mut result = Self::ZERO;
+        let mut iter = s.chars().peekable();
+        while let Some(c) = iter.next() {
+            let n = iter.peek();
+            result.advance(c, n)
         }
-        Self { utf8, utf16, char }
+        result
     }
 }
 
 impl From<&String> for TextIndex {
```

Comment:
It's fine for convenience but I think we should probably get rid of this impl altogether; the only reason we have it is because we don't autoderef String through Rc to &str; see #964

Reply:
I think we shouldn't do that in this PR